specs/stein/approved/amd-sev-libvirt-support.rst

```diff
@@ -0,0 +1,367 @@
+- `manually assign that trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#resource-provider-trait-set>`_
+  to each compute node which is SEV-capable.
+
+This trait would also indicate that the libvirt driver should include
+extra XML in the guest's domain definition, to enable SEV security,
+ensure that the VM uses machine type ``q35`` and UEFI, and that
+``iommu_platform=on`` is used for all virtio devices, e.g.::
+
+    <domain type='kvm'>
+      <os>
+        <type arch='x86_64' machine='pc-q35-2.11'>hvm</type>
+        <loader readonly='yes' type='pflash'>/usr/share/qemu/ovmf-x86_64-ms-4m-code.bin</loader>
+        <nvram>/var/lib/libvirt/qemu/nvram/sles15-sev-guest_VARS.fd</nvram>
+        <boot dev='hd'/>
+      </os>
+      <launchSecurity type='sev'>
+        <cbitpos>47</cbitpos>
+        <reducedPhysBits>1</reducedPhysBits>
+        <policy>0x0001</policy>
+      </launchSecurity>
+      ...
+    </domain>
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
+
+This additional functionality could be implemented within
+``nova.virt.libvirt.LibvirtDriver._set_features()``, e.g. using a new
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
+Immediate follow-up work
+------------------------
+
+It is suggested that follow-up specs be submitted detailing
+the following subsequent work:
+
+- Support a built-in trait which would be automatically detected per
+  compute host by nova's libvirt driver.  This would most likely be
+  called ``HW_CPU_x86_SEV`` or similar, as an extension of `the
+  existing CPU traits mapping
+  <https://github.com/openstack/nova/blob/c5a7002bd571379818c0108296041d12bc171728/nova/virt/libvirt/utils.py#L47>`_.
+
+  Additional logic would be required to check that the various layers
+  of the hardware and software hypervisor stack are *all* SEV-capable:
+
+  - The presence of the following XML in the response from a libvirt
+    `virConnectGetDomainCapabilities()
+    <https://libvirt.org/html/libvirt-libvirt-domain.html#virConnectGetDomainCapabilities>`_
+    API call `indicates that both QEMU and the AMD Secure Processor
+    (AMD-SP) support SEV functionality
+    <https://libvirt.org/git/?p=libvirt.git;a=commit;h=6688393c6b222b5d7cba238f21d55134611ede9c>`_::
+
+        <domainCapabilities>
+          ...
+          <features>
+            ...
+            <sev supported='yes'/>
+              ...
+            </sev>
+          </features>
+        </domainCapabilities>
+
+  - ``/sys/module/kvm_amd/parameters/sev`` should have the value one
+    to indicate that the kernel has SEV capabilities enabled.
+
+- Since `migration of instances between unencrypted and SEV-encrypted
```

Comment:
does it allow to start unencrypted instances on sev enabled hypervisors? aka is this optional or then a required feature, with other words does the hypervisor become dedicated to sEV workload with this change or not?

Reply:
It's optional.